File: src/com/danachury/samples/learningkotlin/java/FunctionalProgrammingExamples.java
package com.danachury.samples.learningkotlin.java;

import java.util.ArrayList;
import java.util.function.Function;

import static java.util.Arrays.asList;

public class FunctionalProgrammingExamples {

    private static Function<String, String> toSentenceCase = x -> x.substring(0, 1).toUpperCase() + x.substring(1);

    private static <T> T applyFunction(T input, Function<T, T> fun) {
        return fun.apply(input);
    }

    public static void main(String[] args) {
        final var result = applyFunction("hello", toSentenceCase);
        System.out.println(result);

        final var colors = asList(
            "red",
            "green",
            "blue",
            "black"
        );

        final var uppercaseColors = new ArrayList<String>();
        for (String color : colors)
            uppercaseColors.add(color.toUpperCase());

        uppercaseColors.forEach(System.out::println);
    }
}

File: src/com/danachury/samples/learningkotlin/kotlin/FunctionalProgrammingExamples.kt
package com.danachury.samples.learningkotlin.kotlin

import com.danachury.samples.learningkotlin.kotlin.util.apply
import com.danachury.samples.learningkotlin.kotlin.util.toSentenceCase
import kotlin.math.max

fun main(args: Array<String>) {
    val result = apply("hello", ::toSentenceCase)
    val result2 = apply("hello") { it.toUpperCase() }
    println(result)
    println(result2)

    val colors = listOf(
        "red",
        "green",
        "blue",
        "black"
    )

    val uppercaseColors = colors.map { it.toUpperCase() }

    uppercaseColors.forEach(::println)

    println("Colors starts with: b")
    colors
        .filter { it.startsWith("b") }
        .forEach(::println)

    println("Duplicate colors that starts with: b")
    colors
        .flatMap { if (it.startsWith("b")) listOf(it, it) else listOf(it) }
        .forEach(::println)

    println("Reduce list to a single String")
    println(colors.reduce { s, value -> "$s, $value" })

    println("Operations over list of numbers")
    val numbers = colors.map { it.length }
    println(numbers.sum())
    println(numbers.average())
    println(numbers.count())

    println(numbers.fold(0) { r, value -> r + value })
    println(numbers.fold(0) { r, value -> max(r, value) })

    println("Using Maps -> Filters -> Destructuring")
    val myMap = mapOf(1 to "one", 2 to "two", 3 to "three")
    myMap
        .filter {(_, v) -> v.startsWith("t") }
        .forEach(::println)
}
